File: src/settings/repo.rs
//! Settings persistence over the `settings` key/value table. User-editable
//! (read + write the whole struct). All paths fail silently to a sensible
//! default — a DB hiccup must never crash or block the UI. Also owns the
//! resolution presets and the one-time `settings.json` import.

use super::Settings;
use rusqlite::Connection;
use std::path::Path;

/// Selectable window resolutions, in display order. Index 2 (1600×900) is
/// the default.
pub const PRESETS: [(u32, u32); 4] = [(800, 600), (1280, 720), (1600, 900), (1920, 1080)];

const DEFAULT_INDEX: i32 = 2;

/// Read a single setting's raw value, or `None` if absent or unreadable.
fn get(conn: &Connection, key: &str) -> Option<String> {
    conn.query_row("SELECT value FROM settings WHERE key = ?1", [key], |r| r.get(0))
        .ok()
}

/// Upsert a single setting. Errors are silently ignored.
fn set(conn: &Connection, key: &str, value: &str) {
    let _ = conn.execute(
        "INSERT INTO settings (key, value) VALUES (?1, ?2) \
         ON CONFLICT(key) DO UPDATE SET value = excluded.value",
        rusqlite::params![key, value],
    );
}

/// Load settings from the DB, falling back to defaults for any missing or
/// unreadable key.
pub fn load(conn: &Connection) -> Settings {
    let d = Settings::default();
    Settings {
        width: get(conn, "width").and_then(|v| v.parse().ok()).unwrap_or(d.width),
        height: get(conn, "height").and_then(|v| v.parse().ok()).unwrap_or(d.height),
        dark: get(conn, "dark").map(|v| v == "1").unwrap_or(d.dark),
        fullscreen: get(conn, "fullscreen").map(|v| v == "1").unwrap_or(d.fullscreen),
    }
}

/// Persist all settings fields to the DB. Write errors are silently ignored.
pub fn save(conn: &Connection, settings: &Settings) {
    set(conn, "width", &settings.width.to_string());
    set(conn, "height", &settings.height.to_string());
    set(conn, "dark", if settings.dark { "1" } else { "0" });
    set(conn, "fullscreen", if settings.fullscreen { "1" } else { "0" });
}

/// One-time migration of a pre-SQLite `settings.json` into the DB. If the
/// file exists and parses, its values are persisted and the file is deleted
/// so this never runs again. A missing or corrupt file is left untouched.
pub fn import_legacy(conn: &Connection, json_path: &Path) {
    if !json_path.exists() {
        return;
    }
    let parsed = std::fs::read_to_string(json_path)
        .ok()
        .and_then(|s| serde_json::from_str::<Settings>(&s).ok());
    if let Some(s) = parsed {
        save(conn, &s);
        let _ = std::fs::remove_file(json_path);
    }
}

/// Index into [`PRESETS`] matching the given size, or the default index
/// when no preset matches.
pub fn preset_index(width: u32, height: u32) -> i32 {
    PRESETS
        .iter()
        .position(|&(w, h)| w == width && h == height)
        .map(|i| i as i32)
        .unwrap_or(DEFAULT_INDEX)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A migrated, empty in-memory database.
    fn db() -> Connection {
        let c = Connection::open_in_memory().unwrap();
        crate::db::run_migrations(&c).unwrap();
        c
    }

    #[test]
    fn preset_index_matches_known_size() {
        assert_eq!(preset_index(800, 600), 0);
        assert_eq!(preset_index(1920, 1080), 3);
    }

    #[test]
    fn preset_index_falls_back_for_unknown_size() {
        assert_eq!(preset_index(1234, 567), DEFAULT_INDEX);
    }

    #[test]
    fn load_returns_defaults_on_empty_db() {
        let s = load(&db());
        assert_eq!((s.width, s.height), (1600, 900));
        assert!(s.dark);
        assert!(!s.fullscreen);
    }

    #[test]
    fn load_uses_defaults_for_missing_keys() {
        // Only width/height persisted; theme/fullscreen must fall back.
        let c = db();
        set(&c, "width", "800");
        set(&c, "height", "600");
        let s = load(&c);
        assert_eq!((s.width, s.height), (800, 600));
        assert!(s.dark); // default_dark
        assert!(!s.fullscreen);
    }

    #[test]
    fn save_then_load_roundtrips_all_fields() {
        let c = db();
        save(
            &c,
            &Settings { width: 1280, height: 720, dark: false, fullscreen: true },
        );
        let back = load(&c);
        assert_eq!((back.width, back.height), (1280, 720));
        assert!(!back.dark);
        assert!(back.fullscreen);
    }

    #[test]
    fn import_writes_settings_and_deletes_file() {
        let c = db();
        let path = std::env::temp_dir().join("denso_import_ok.json");
        std::fs::write(
            &path,
            r#"{"width":1280,"height":720,"dark":false,"fullscreen":true}"#,
        )
        .unwrap();

        import_legacy(&c, &path);

        let s = load(&c);
        assert_eq!((s.width, s.height), (1280, 720));
        assert!(!s.dark);
        assert!(s.fullscreen);
        assert!(!path.exists(), "legacy file should be deleted after import");
    }

    #[test]
    fn import_is_noop_when_file_absent() {
        let c = db();
        let path = std::env::temp_dir().join("denso_import_absent.json");
        let _ = std::fs::remove_file(&path);

        import_legacy(&c, &path); // must not panic

        let s = load(&c);
        assert_eq!((s.width, s.height), (1600, 900)); // untouched defaults
    }

    #[test]
    fn import_leaves_corrupt_file_intact() {
        let c = db();
        let path = std::env::temp_dir().join("denso_import_corrupt.json");
        std::fs::write(&path, "}{ not json").unwrap();

        import_legacy(&c, &path);

        assert!(path.exists(), "corrupt file kept for inspection");
        let s = load(&c);
        assert_eq!((s.width, s.height), (1600, 900)); // defaults, nothing imported
        let _ = std::fs::remove_file(&path);
    }

    #[test]
    fn save_overwrites_previous_values() {
        let c = db();
        save(&c, &Settings { width: 800, height: 600, dark: true, fullscreen: false });
        save(&c, &Settings { width: 1920, height: 1080, dark: false, fullscreen: true });
        let back = load(&c);
        assert_eq!((back.width, back.height), (1920, 1080));
        assert!(!back.dark);
        assert!(back.fullscreen);
    }
}

File: src/settings/repo.cpp
#include "settings/repo.h"

#include <QByteArray>
#include <QFile>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonParseError>
#include <QJsonValue>
#include <QSqlQuery>
#include <QVariant>

#include <optional>

namespace denso::settings {

namespace {

/// Read a single setting's raw value, or nullopt if absent or unreadable.
std::optional<QString> get(const QSqlDatabase& db, const QString& key) {
    QSqlQuery q(db);
    q.prepare(QStringLiteral("SELECT value FROM settings WHERE key = ?"));
    q.addBindValue(key);
    if (q.exec() && q.next()) {
        return q.value(0).toString();
    }
    return std::nullopt;
}

/// Upsert a single setting. Errors are silently ignored.
void set(const QSqlDatabase& db, const QString& key, const QString& value) {
    QSqlQuery q(db);
    q.prepare(QStringLiteral(
        "INSERT INTO settings (key, value) VALUES (?, ?) "
        "ON CONFLICT(key) DO UPDATE SET value = excluded.value"));
    q.addBindValue(key);
    q.addBindValue(value);
    q.exec();
}

} // namespace

Settings load(const QSqlDatabase& db) {
    const Settings d;  // defaults
    Settings out = d;

    if (const auto v = get(db, QStringLiteral("width"))) {
        bool ok = false;
        const uint parsed = v->toUInt(&ok);
        if (ok) {
            out.width = parsed;
        }
    }
    if (const auto v = get(db, QStringLiteral("height"))) {
        bool ok = false;
        const uint parsed = v->toUInt(&ok);
        if (ok) {
            out.height = parsed;
        }
    }
    if (const auto v = get(db, QStringLiteral("dark"))) {
        out.dark = (*v == QStringLiteral("1"));
    }
    if (const auto v = get(db, QStringLiteral("fullscreen"))) {
        out.fullscreen = (*v == QStringLiteral("1"));
    }
    return out;
}

void save(const QSqlDatabase& db, const Settings& settings) {
    set(db, QStringLiteral("width"), QString::number(settings.width));
    set(db, QStringLiteral("height"), QString::number(settings.height));
    set(db, QStringLiteral("dark"), settings.dark ? QStringLiteral("1") : QStringLiteral("0"));
    set(db, QStringLiteral("fullscreen"),
        settings.fullscreen ? QStringLiteral("1") : QStringLiteral("0"));
}

void import_legacy(const QSqlDatabase& db, const QString& json_path) {
    if (!QFile::exists(json_path)) {
        return;
    }
    QFile file(json_path);
    if (!file.open(QIODevice::ReadOnly)) {
        return;  // unreadable → leave untouched (mirrors read_to_string().ok())
    }
    const QByteArray bytes = file.readAll();
    file.close();

    QJsonParseError err{};
    const QJsonDocument doc = QJsonDocument::fromJson(bytes, &err);
    if (err.error != QJsonParseError::NoError || !doc.isObject()) {
        return;  // corrupt → kept for inspection, nothing imported
    }
    const QJsonObject obj = doc.object();

    // serde requires width/height (no default) and rejects wrong-typed fields;
    // dark defaults to true, fullscreen to false. Mirror that contract exactly.
    // serde's u32 also rejects negative, fractional, or out-of-range numbers —
    // a plain isDouble() would wrongly accept "-5"/"3.5"/"5e12", so require an
    // exact, in-range integer here too (all valid u32 are exact as double).
    const QJsonValue wv = obj.value(QStringLiteral("width"));
    const QJsonValue hv = obj.value(QStringLiteral("height"));
    const auto is_u32 = [](const QJsonValue& v) {
        if (!v.isDouble()) {
            return false;
        }
        const double d = v.toDouble();
        return d >= 0.0 && d <= 4294967295.0
            && d == static_cast<double>(static_cast<uint32_t>(d));
    };
    if (!is_u32(wv) || !is_u32(hv)) {
        return;
    }
    Settings s;  // defaults supply dark=true, fullscreen=false
    s.width = static_cast<uint32_t>(wv.toDouble());
    s.height = static_cast<uint32_t>(hv.toDouble());
    if (obj.contains(QStringLiteral("dark"))) {
        const QJsonValue dv = obj.value(QStringLiteral("dark"));
        if (!dv.isBool()) {
            return;
        }
        s.dark = dv.toBool();
    }
    if (obj.contains(QStringLiteral("fullscreen"))) {
        const QJsonValue fv = obj.value(QStringLiteral("fullscreen"));
        if (!fv.isBool()) {
            return;
        }
        s.fullscreen = fv.toBool();
    }

    save(db, s);
    QFile::remove(json_path);
}

} // namespace denso::settings
